presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+                ranges.add((ParquetRangeStatistics<T>) statistics);
+            }
+            else if (min instanceof Double) {
+                ParquetDoubleStatistics statistics = new ParquetDoubleStatistics((Double) min, (Double) max);
+                ranges.add((ParquetRangeStatistics<T>) statistics);
+            }
+            else if (min instanceof Slice) {
+                ParquetStringStatistics statistics = new ParquetStringStatistics((Slice) min, (Slice) max);
+                ranges.add((ParquetRangeStatistics<T>) statistics);
+            }
+        }
+        return createDomain(type, hasNullValue, ranges);
+    }
+
+    private boolean isCorruptedColumnIndex(ColumnIndex columnIndex)
+    {
+        if (columnIndex.getMaxValues() == null || columnIndex.getMinValues() == null ||
+                columnIndex.getNullCounts() == null || columnIndex.getNullPages() == null) {
+            return true;
+        }
+
+        if (columnIndex.getMaxValues().size() != columnIndex.getMinValues().size() ||
+                columnIndex.getMaxValues().size() != columnIndex.getNullPages().size() ||
+                columnIndex.getMaxValues().size() != columnIndex.getNullCounts().size()) {
+            return true;
+        }
+
+        return false;
+    }
+
+    // Caller should verify isCorruptedColumnIndex is false first
+    private boolean isEmptyColumnIndex(ColumnIndex columnIndex)
+    {
+        return columnIndex.getMaxValues().size() == 0;
+    }
+
+    public FilterPredicate convertToParquetUdp()
+    {
+        FilterPredicate filter = null;
+
+        // TODO: It could be a Presto bug that we don't see effectivePredicate.getDomains().get() has more than 1 domain.
+        //  For example, the 'where c1=3 or c1=10002' clause should have two domains but it has none
+        //  Todo: we assume the relation cross domains are 'or'
+        for (RichColumnDescriptor column : columns) {
+            Domain domain = effectivePredicate.getDomains().get().get(column);
+            if (domain == null || domain.isNone()) {
+                continue;
+            }
+
+            if (domain.isAll()) {
+                continue;
+            }
+
+            FilterPredicate columnFilter = FilterApi.userDefined(FilterApi.intColumn(ColumnPath.get(column.getPath()).toDotString()), new DomainUserDefinedPredicate(domain));
+            if (filter == null) {
+                filter = columnFilter;
+            }
+            else {
+                filter = FilterApi.or(filter, columnFilter);
+            }
+        }
+
+        return filter;
+    }
+
+    /**
+     * This class implements methods defined in UserDefinedPredicate based on the page statistic and tuple domain(for a column).
+     */
+    static class DomainUserDefinedPredicate<T extends Comparable<T>>
+            extends UserDefinedPredicate<T>
+            implements Serializable
```

Comment:
Why Serializable?

Reply:
Being Serializable is required by argument of FilterApi.userDefined call below